src/registrar/tests/test_models_domain.py

```diff
@@ -22,6 +59,70 @@ def test_minimal_create(self):
         # above command didn't error out is proof enough
         # self.assertEquals(domain.state, Domain.State.DRAFTED)
 
+    def test_cache_sets_resets(self):
+        """Cache should be set on getter and reset on setter calls"""
+        domain, _ = Domain.objects.get_or_create(name="igorville.gov")
+        # trigger getter
+        _ = domain.creation_date
+
+        # getter should set the domain cache with a InfoDomain object
+        # (see InfoDomainResult)
+        self.assertEquals(domain._cache["auth_info"], self.mockDataInfoDomain.auth_info)
+        self.assertEquals(domain._cache["cr_date"], self.mockDataInfoDomain.cr_date)
+        self.assertFalse("avail" in domain._cache.keys())
+
+        # using a setter should clear the cache
+        domain.nameservers = [("", "")]
+        self.assertEquals(domain._cache, {})
+
+        # send should have been called only once
+        self.mock_foo.assert_called_once()
+
+    def test_cache_used_when_avail(self):
+        """Cache is pulled from if the object has already been accessed"""
+        domain, _ = Domain.objects.get_or_create(name="igorville.gov")
+        cr_date = domain.creation_date
+
+        # repeat the getter call
+        cr_date = domain.creation_date
+
+        # value should still be set correctly
+        self.assertEqual(cr_date, self.mockDataInfoDomain.cr_date)
+        self.assertEqual(domain._cache["cr_date"], self.mockDataInfoDomain.cr_date)
+
+        # send was only called once & not on the second getter call
+        self.mock_foo.assert_called_once()
+
+    def test_cache_nested_elements(self):
+        """Cache works correctly with the nested objects cache and hosts"""
+        domain, _ = Domain.objects.get_or_create(name="igorville.gov")
+
+        # the cached contacts and hosts should be dictionaries of what is passed to them
+        expectedContactsDict = {
+            "id": self.mockDataInfoDomain.contacts[0],
+            "auth_info": self.mockDataInfoContact.auth_info,
+            "cr_date": self.mockDataInfoContact.cr_date,
+        }
+        expectedHostsDict = {
+            "name": self.mockDataInfoDomain.hosts[0],
+            "cr_date": self.mockDataInfoHosts.cr_date,
+        }
+
+        # this can be changed when the getter for contacts is implemented
+        domain._get_property("contacts")
+
+        # check domain info is still correct and not overridden
+        self.assertEqual(domain._cache["auth_info"], self.mockDataInfoDomain.auth_info)
+        self.assertEqual(domain._cache["cr_date"], self.mockDataInfoDomain.cr_date)
+
+        # check contacts
```

Comment:
**comment (non-blocking)**: This seems non-obvious that there are two "contacts" things under different keys. Could we have more explanation in a comment here about why, or refer to a docstring about it?

Reply:
Good point there for sure be a docstring on it and comments! i will do that